### PR TITLE
Reward API was added

### PR DESCRIPTION
**Pull request by ahmedrns**
Validator reward API was added
Block stat method was updated
getAccountInfo method was updated
save_stats method was updated

### CODE REVIEW

File: src/Types.ts [hunk 1]
```diff
@@ -806,6 +806,41 @@ export interface IAccountInformation {
...
+     */
+    block_reward: number;
+
+    /**
+     * Total block fee
+     */
+    block_fee: number;
+
+    /**
+     * Validator receive reward
```
Comment (TrustHenry) [inline src/Types.ts]:
```suggestion
     * Validator reward amount
```

File: src/Stoa.ts [hunk 2]
```diff
@@ -2580,6 +2582,51 @@ class Stoa extends WebService {
...
+            .getValidatorReward(address, pagination.pageSize, pagination.page)
+            .then((data: any[]) => {
+                if (data.length === 0) {
+                    return res.status(204).send(`The data does not exist.`);
+                } else {
+                    let rewards: IValidatorReward[] = []
+                    for (const row of data) {
+                        rewards.push({
+                            block_height: row.block_height,
+                            freezing_amount: row.type === 1 ? row.amount : 0,
```
Comment (TrustHenry) [inline src/Stoa.ts]: Freezing_amount is actually the amount that the node has staked.
This is the UTXO amount of enrollment.

File: src/Types.ts [hunk 3]
```diff
@@ -806,6 +806,41 @@ export interface IAccountInformation {
...
+ * Interface for Validator Reward
+ */
+export interface IValidatorReward {
+    /**
+     * Block height
+     */
+    block_height: number;
+
+    /**
+     * Freezing amount
```
Comment (TrustHenry) [inline src/Types.ts]: The amount of steaking

Comment (ahmedrns) [reply src/Types.ts]: resolved

File: src/Types.ts (hunk 1 above)
Comment (ahmedrns) [reply src/Types.ts]: resolved

File: src/Stoa.ts (hunk 2 above)
Comment (ahmedrns) [reply src/Stoa.ts]: validator staked amount will be displayed

File: tests/Boascan.test.ts [hunk 4]
```diff
@@ -1046,4 +1046,12 @@ describe("Test of Stoa API Server", () => {
         }
         assert.deepStrictEqual(response.data, expected);
     });
+
+    it("Test for path /validator/reward/:address", async () => {
+        const uri = URI(stoa_addr)
+            .directory("/validator/reward")
+            .filename("boa1xpvald2ydpxzl9aat978kv78y5g24jxy46mcnl7munf4jyhd0zjrc5x62kn");
+        const response = await client.get(uri.toString());
+        assert.ok(response);
```
Comment (TrustHenry) [inline tests/Boascan.test.ts]: block_height, steaking_amount, block_reward, block_fee, validator_reward, total_count
Test values are not being tested.

File: src/modules/storage/LedgerStorage.ts [hunk 5]
```diff
@@ -3696,6 +3711,34 @@ export class LedgerStorage extends Storages {
...
+                O.amount as validator_reward,
+                B.total_reward,
+                B.total_fee,
+                count(*) OVER() AS full_count
+                FROM validators V
+                INNER JOIN tx_outputs O ON(V.address = O.address)
+                INNER JOIN blocks_stats B ON (O.block_height = B.block_height)
+                WHERE 
+                (V.address = ? AND 
+                V.enrolled_at = (select max(enrolled_at) from validators WHERE address = ?)
```
Comment (TrustHenry) [inline src/modules/storage/LedgerStorage.ts]: It must be connected to a valid enrollment at the corresponding block height.

Comment (ahmedrns) [reply src/modules/storage/LedgerStorage.ts]: When the validator gets reward, validator's latest enrollment will be fetched

Comment (TrustHenry) [reply src/modules/storage/LedgerStorage.ts]: It is also inquired at the height of the past block.

File: src/modules/storage/LedgerStorage.ts [hunk 6]
```diff
@@ -3696,6 +3711,37 @@ export class LedgerStorage extends Storages {
...
+                count(*) OVER() AS full_count
+                FROM tx_outputs O
+                LEFT JOIN validators V ON(V.address = O.address)
+                INNER JOIN blocks_stats B ON (O.block_height = B.block_height)
+                WHERE 
+                (O.address = ? 
+                AND 
+                (IFNULL(V.enrolled_at,B.block_height + ${this.validator_cycle}) >= (B.block_height - ${this.validator_cycle}) 
+                AND (IFNULL(V.enrolled_at,(B.block_height - 1)) < B.block_height))
+                AND O.type = 2)
```
Comment (TrustHenry) [inline src/modules/storage/LedgerStorage.ts]:
```suggestion
                LEFT OUTER JOIN validators V ON
                    V.address = O.address
                    AND V.enrolled_at >= (O.block_height - ${this.validator_cycle})
                    AND V.enrolled_at < O.block_height
                INNER JOIN blocks_stats B ON 
                    O.block_height = B.block_height
                WHERE 
                    O.address = ? 
                    AND O.type = 2
```

File: src/modules/storage/LedgerStorage.ts [hunk 7]
```diff
@@ -3696,6 +3711,37 @@ export class LedgerStorage extends Storages {
...
+                LIMIT ? OFFSET ?`;
```
Comment (TrustHenry) [inline src/modules/storage/LedgerStorage.ts:3743]:
```suggestion
                LIMIT ? OFFSET ?
                ORDER BY O.block_height ASC`;
```

**Review by TrustHenry — APPROVED** [review]
LGTM

**Pull request merged**